File: wallet-android/chapter3/code/app/src/main/java/com/digquant/util/DensityUtil.java
package com.digquant.util;

import android.content.Context;
import android.util.TypedValue;

public class DensityUtil {



    public static int DP2PX(Context context, float dp) {
        final float scale = context.getResources().getDisplayMetrics().density;

        return (int) (dp * scale + 0.5f);
    }

    public static int PX2DP(Context context, float px) {

        final float scale = context.getResources().getDisplayMetrics().density;

        return (int) (px / scale + 0.5f);
    }

    public static int SP2PX(Context context, float sp) {

        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, context.getResources().getDisplayMetrics());
    }
}

File: wallet-android/chapter8/code/app/src/main/java/com/digquant/util/DensityUtil.kt
package com.digquant.util

import android.content.Context
import android.util.TypedValue

object DensityUtil {
    fun DP2PX(context: Context, dp: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dp * scale + 0.5f).toInt()
    }

    fun PX2DP(context: Context, px: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (px / scale + 0.5f).toInt()
    }

    fun SP2PX(context: Context, sp: Float): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP,
            sp,
            context.resources.displayMetrics
        )
            .toInt()
    }
}
